gnucash/gnome-utils/dialog-options.cpp

```diff
@@ -0,0 +1,2829 @@
+
+    /* Hold onto this for a complete reset */
+    GncOptionDB *option_db;
+
+    /* Hold on to this to unregister the right class */
+    const char *component_class;
+
+    /* widget being destroyed */
+    bool destroyed;
+};
+
+enum page_tree
+{
+    PAGE_INDEX = 0,
+    PAGE_NAME,
+    NUM_COLUMNS
+};
+
+//Init the class static.
+std::vector<WidgetCreateFunc> GncOptionUIFactory::s_registry{static_cast<size_t>(GncOptionUIType::MAX_VALUE)};
+
+void
+GncOptionUIFactory::set_func(GncOptionUIType type, WidgetCreateFunc func)
+{
+    s_registry[static_cast<size_t>(type)] = func;
+}
+
+GtkWidget*
+GncOptionUIFactory::create(GncOption& option, GtkGrid* page, GtkLabel* name,
+                     char* description, GtkWidget** enclosing, bool* packed)
+{
+    auto type{option.get_ui_type()};
+    auto func{s_registry[static_cast<size_t>(type)]};
+    if (func)
+        return func(option, page, name, description, enclosing, packed);
+    PERR("No function registered for type %d", static_cast<int>(type));
+    return nullptr;
+}
+
+GncOptionGtkUIItem::GncOptionGtkUIItem(GtkWidget* widget,
+                                       GncOptionUIType type) :
+    GncOptionUIItem{type},
+    m_widget{static_cast<GtkWidget*>(g_object_ref(widget))} {}
+
+GncOptionGtkUIItem::GncOptionGtkUIItem(const GncOptionGtkUIItem& item) :
+        GncOptionUIItem{item.get_ui_type()},
+        m_widget{static_cast<GtkWidget*>(g_object_ref(item.get_widget()))} {}
+
+GncOptionGtkUIItem::~GncOptionGtkUIItem()
+{
+    if (m_widget)
+        g_object_unref(m_widget);
+}
+
+void
+GncOptionGtkUIItem::set_selectable(bool selectable) const noexcept
+{
+    if (m_widget)
+        gtk_widget_set_sensitive (m_widget, selectable);
+}
+
+void
+GncOptionGtkUIItem::clear_ui_item()
+{
+    if (m_widget)
+        g_object_unref(m_widget);
+    m_widget = nullptr;
+}
+
+void
+GncOptionGtkUIItem::set_widget(GtkWidget* widget)
+{
+    if (get_ui_type() == GncOptionUIType::INTERNAL)
+    {
+        std::string error{"INTERNAL option, setting the UI item forbidden."};
+        throw std::logic_error(std::move(error));
+    }
+
+    if (m_widget)
+        g_object_unref(m_widget);
+
+    m_widget = static_cast<GtkWidget*>(g_object_ref(widget));
+}
+
+
+static GNCOptionWinCallback global_help_cb = NULL;
+gpointer global_help_cb_data = NULL;
+
+static void dialog_reset_cb(GtkWidget * w, gpointer data);
+void dialog_list_select_cb (GtkTreeSelection *selection,
+                                        gpointer data);
+static void component_close_handler (gpointer data);
+
+static void
+section_reset_widgets(GncOptionSection* section)
+{
+}
+
+GtkWidget* const
+gnc_option_get_gtk_widget (const GncOption* option)
+{
+    if (!option) return nullptr;
+    auto ui_item{dynamic_cast<const GncOptionGtkUIItem*>(option->get_ui_item())};
+    if (ui_item)
+        return ui_item->get_widget();
+
+    return nullptr;
+}
+
+static void
+dialog_changed_internal (GtkWidget *widget, bool sensitive)
```

Comment:
@jralls 
Opened the balance sheet report and then options and made a change but 'Apply and OK not active'. I changed this function to this which enables the buttons but unfortunately this  has them enabled on startup. Previously the last call to this function on startup was a call from gnc_options_dialog_build_contents_full with sensitivity set to FALSE but on looking now there are a couple of calls from a GtkBox widget that trace back to the GNCDateEdit option. Not sure if this is related but I was also getting this error...
sys:1: Warning: g_object_ref: assertion 'G_IS_OBJECT (object)' failed

which I think is coming from...
GncOptionDateUIItem::set_ui_item_from_option(GncOption&)



    g_return_if_fail(widget);
    
    if (!GTK_IS_WIDGET(widget))
        return;

    while (TRUE)
    {
        auto new_widget = gtk_widget_get_parent(widget);
        if (new_widget && GTK_IS_WIDGET(new_widget) &&
            !GTK_IS_WINDOW(new_widget))
            widget = new_widget;
        else
            break;
    }

    /* find the ok and cancel buttons, we know where they will be so do it
       this way as opposed to using gtk_container_foreach, much less iteration */
    if (GTK_IS_CONTAINER(widget))
    {
        GList *children = gtk_container_get_children(GTK_CONTAINER(widget));
        for (GList *it = children; it; it = it->next)
        {
            if (GTK_IS_BUTTON_BOX (GTK_WIDGET(it->data)))
            {
                GList *children = gtk_container_get_children(GTK_CONTAINER(it->data));
                for (GList *it = children; it; it = it->next)
                {
                    GtkWidget* widget = GTK_WIDGET(it->data);
                    const gchar* name = gtk_widget_get_name(widget);

                    if (g_strcmp0 (name, "ok_button") == 0 ||
                        g_strcmp0 (name, "apply_button") == 0)
                            gtk_widget_set_sensitive (widget, sensitive);
                    else if (g_strcmp0 (name, "cancel_button") == 0)
                        gtk_button_set_label (GTK_BUTTON (widget),
                                              sensitive ? _("_Cancel") :
                                                          _("_Close"));
                }
                g_list_free (children);
                break; // Found the button-box, no need to continue.
            }
        }
        g_list_free (children);
    }

Reply:
@christopherlam these things don't have history. What did you change?

Reply:
Don't think this is my change.

I can confirm @Bob-IT 's report; usually modifying options would change Report Options' buttons sensitivity from only "Close" enabled, to "Cancel" "Apply" "OK".

I attach trace output

```
sys:1: Warning: g_object_ref: assertion 'G_IS_OBJECT (object)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:54:51 ERROR <Gtk> gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed
* 18:55:10  WARN <gnc.app-utils> [gnc_option_db_destroy()] Direct Destroy called on GncOptionDB 93951560704400
```

as well as gdb's output

```
#0  0x00007ffff7b5ed97 in g_logv () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#1  0x00007ffff7b5f053 in g_log () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#2  0x00007ffff6d185a0 in g_object_ref () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#3  0x00007ffff77e67f7 in GncOptionGtkUIItem::GncOptionGtkUIItem(_GtkWidget*, GncOptionUIType) (this=0x5555565c4c70, widget=0x0, type=GncOptionUIType::DATE_BOTH)
    at /home/chris/sources/gnucash/master/gnucash/gnome-utils/dialog-options.cpp:161
#4  0x00007ffff77f0fac in GncOptionDateUIItem::GncOptionDateUIItem(std::unique_ptr<GncDateEntry, std::default_delete<GncDateEntry> >, GncOptionUIType)
    (this=0x5555565c4c70, entry=std::unique_ptr<GncDateEntry> = {...}, type=GncOptionUIType::DATE_BOTH)
    at /home/chris/sources/gnucash/master/gnucash/gnome-utils/dialog-options.cpp:1430
#5  0x00007ffff77f44c5 in std::make_unique<GncOptionDateUIItem, std::unique_ptr<BothDateEntry, std::default_delete<BothDateEntry> >, GncOptionUIType&>(std::unique_ptr<BothDateEntry, std::default_delete<BothDateEntry> >&&, GncOptionUIType&) () at /usr/include/c++/11/bits/unique_ptr.h:962
#6  0x00007ffff77eac0d in create_date_option_widget(GncOption&, GtkGrid*, GtkLabel*, char*, GtkWidget**, bool*) (option=..., page_box=
    0x555556568b50, name_label=0x55555657fd10, documentation=0x5555563781f0 "Select a date to report on.", enclosing=0x7fffffffb410, packed=0x7fffffffb407)
    at /home/chris/sources/gnucash/master/gnucash/gnome-utils/dialog-options.cpp:1494
#7  0x00007ffff77eb02f in create_option_widget<(GncOptionUIType)9>(GncOption&, _GtkGrid*, _GtkLabel*, char*, _GtkWidget**, bool*)
    (option=..., page_box=0x555556568b50, name_label=0x55555657fd10, documentation=0x5555563781f0 "Select a date to report on.", enclosing=0x7fffffffb410, packed=0x7fffffffb407)
    at /home/chris/sources/gnucash/master/gnucash/gnome-utils/dialog-options.cpp:1571
#8  0x00007ffff77e6767 in GncOptionUIFactory::create(GncOption&, _GtkGrid*, _GtkLabel*, char*, _GtkWidget**, bool*)
    (option=..., page=0x555556568b50, name=0x55555657fd10, description=0x5555563781f0 "Select a date to report on.", enclosing=0x7fffffffb410, packed=0x7fffffffb407)
    at /home/chris/sources/gnucash/master/gnucash/gnome-utils/dialog-options.cpp:153
#9  0x00007ffff77e73ce in gnc_option_set_ui_widget(GncOption&, GtkGrid*, gint) (option=..., page_box=0x555556568b50, grid_row=4)
    at /home/chris/sources/gnucash/master/gnucash/gnome-utils/dialog-options.cpp:348
#10 0x00007ffff77e7b6c in operator()(GncOption&) const (__closure=0x7fffffffb5c0, option=...) at /home/chris/sources/gnucash/master/gnucash/gnome-utils/dialog-options.cpp:486
#11 0x00007ffff77efbcd in std::__invoke_impl<void, gnc_options_dialog_append_page(GNCOptionWin*, GncOptionSectionPtr&)::<lambda(GncOption&)>&, GncOption&>(std::__invoke_other, struct {...} &) (__f=...) at /usr/include/c++/11/bits/invoke.h:61
#12 0x00007ffff77ef521 in std::__invoke_r<void, gnc_options_dialog_append_page(GNCOptionWin*, GncOptionSectionPtr&)::<lambda(GncOption&)>&, GncOption&>(struct {...} &) (__fn=...)
    at /usr/include/c++/11/bits/invoke.h:111
#13 0x00007ffff77eefad in std::_Function_handler<void(GncOption&), gnc_options_dialog_append_page(GNCOptionWin*, GncOptionSectionPtr&)::<lambda(GncOption&)> >::_M_invoke(const std::_Any_data &, GncOption &) (__functor=..., __args#0=...) at /usr/include/c++/11/bits/std_function.h:291
#14 0x00007ffff6bfb15d in std::function<void (GncOption&)>::operator()(GncOption&) const (this=0x7fffffffb5c0, __args#0=...) at /usr/include/c++/11/bits/std_function.h:560
#15 0x00007ffff6bf9058 in std::for_each<__gnu_cxx::__normal_iterator<GncOption*, std::vector<GncOption, std::allocator<GncOption> > >, std::function<void (GncOption&)> >(__gnu_cxx::__normal_iterator<GncOption*, std::vector<GncOption, std::allocator<GncOption> > >, __gnu_cxx::__normal_iterator<GncOption*, std::vector<GncOption, std::allocator<GncOption> > >, std::function<void (GncOption&)>) (__first=
    {static c_empty_string = {static npos = 18446744073709551615, _M_dataplus = {<std::allocator<char>> = {<__gnu_cxx::new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0x7ffff7affdd0 <GncOption::c_empty_string[abi:cxx11]+16> ""}, _M_string_length = 0, {_M_local_buf = '\000' <repeats 15 times>, _M_allocated_capacity = 0}}, m_option = std::unique_ptr<std::variant<GncOptionValue<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, GncOptionValue<bool>, GncOptionValue<long>, GncOptionValue<QofInstance_s const*>, GncOptionValue<_QofQuery const*>, GncOptionValue<_gncOwner const*>, GncOptionValue<scm_unused_struct*>, GncOptionAccountListValue, GncOptionAccountSelValue, GncOptionMultichoiceValue, GncOptionRangeValue<int>, GncOptionRangeValue<double>, GncOptionValidatedValue<QofInstance_s const*>, GncOptionValidatedValue<_QofQuery const*>, GncOptionDateValue>> = {get() = {<No data fields>}}, m_ui_item = std::unique_ptr<GncOptionUIItem> = {get() = {<No data fields>}}}, __last=
    {static c_empty_string = {static npos = 18446744073709551615, _M_dataplus = {<std::allocator<char>> = {<__gnu_cxx::new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0x7ffff7affdd0 <GncOption::c_empty_string[abi:cxx11]+16> ""}, _M_string_length = 0, {_M_local_buf = '\000' <repeats 15 times>, _M_allocated_capacity = 0}}, m_option = std::unique_ptr<std::variant<GncOptionValue<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, GncOptionValue<bool>, GncOptionValue<long>, GncOptionValue<QofInstance_s const*>, GncOptionValue<_QofQuery const*>, GncOptionValue<_gncOwner const*>, GncOptionValue<scm_unused_struct*>, GncOptionAccountListValue, GncOptionAccountSelValue, GncOptionMultichoiceValue, GncOptionRangeValue<int>, GncOptionRangeValue<double>, GncOptionValidatedValue<QofInstance_s const*>, GncOptionValidatedValue<_QofQuery const*>, GncOptionDateValue>> = {get() = {<No data fields>}}, m_ui_item = std::unique_ptr<GncOptionUIItem> = {get() = {<No data fields>}}}, __f=...) at /usr/include/c++/11/bits/stl_algo.h:3820
#16 0x00007ffff6bec7db in GncOptionSection::foreach_option(std::function<void (GncOption&)>) (this=0x555556378db0, func=...)
    at /home/chris/sources/gnucash/master/libgnucash/app-utils/gnc-optiondb.cpp:123
#17 0x00007ffff77e7d13 in gnc_options_dialog_append_page(GNCOptionWin*, GncOptionSectionPtr&)
    (propertybox=0x5555559f4fb0, section=std::shared_ptr<GncOptionSection> (use count 1, weak count 0) = {...})
--Type <RET> for more, q to quit, c to continue without paging--q
Quit
(gdb) f 3
#3  0x00007ffff77e67f7 in GncOptionGtkUIItem::GncOptionGtkUIItem (this=0x5555565c4c70, widget=0x0, type=GncOptionUIType::DATE_BOTH)
    at /home/chris/sources/gnucash/master/gnucash/gnome-utils/dialog-options.cpp:161
161	    m_widget{static_cast<GtkWidget*>(g_object_ref(widget))} {}
(gdb) p widget
$1 = 0x0
(gdb) 
```

Reply:
@christopherlam 

> Don't think this is my change.

Github tagged @Bob-IT's original comment with *edited by Christopher Lam*. What did you change in Bob's comment?

Reply:
@jralls 
I think @christopherlam  might of added code blocks to make it look better. If you want to know what I basically changed I replaced the function with the code above which does the following...
Added the GTK_IS_WIDGET at the start to remove the parent failures.
In the while loop, changed to !GTK_IS_WINDOW so widget is the vertical box below the window and so adjusted the if statements below to remove one level of testing.

Reply:
@Bob-IT Ok.

ISTM it's an error to call `dialog_changed_internal` with a `widget` that isn't a `GtkWidget*`. I'll track that down.

Reply:
> @jralls I think @christopherlam might of added code blocks to make it look better. If you want to know what I basically changed I replaced the function with the code above which does the following... Added the GTK_IS_WIDGET at the start to remove the parent failures. In the while loop, changed to !GTK_IS_WINDOW so widget is the vertical box below the window and so adjusted the if statements below to remove one level of testing.

I'd changed \`\`\` to \`\`\`c; I need fisher-price alphabet to read code in emacs...

(I didn't mean to edit bob's reply, I meant to reply to it :fearful:

Reply:
> @jralls I think @christopherlam might of added code blocks to make it look better. If you want to know what I basically changed I replaced the function with the code above which does the following... Added the GTK_IS_WIDGET at the start to remove the parent failures. In the while loop, changed to !GTK_IS_WINDOW so widget is the vertical box below the window and so adjusted the if statements below to remove one level of testing.

We have two common functions in `gnucash/gnome-utils/gnc-ui.h` to do exactly that: it finds the GtkWindow or GtkDialog that holds `widget`:
* `gnc_ui_get_gtk_window`: finds the immediate GtkWindow or GtkDialog and stops there
* `gnc_ui_get_main_window: starts by finding the immediate GtkWindow or GtkDialog. If it's a GtkDialog the code will continue to determine the final GtkWindow this dialog is attached to.
I suspect what you need is the first of these two.

Reply:
@gjanssens, Gtk has a function for the first, `gtk_widget_get_toplevel`. All of them depend on the widget you start from and all the containers up to the toplevel being correctly parented, and some weren't.

But @Bob-IT was anyway trying to get the top GtkVBox that the window contains. Unfortunately that didn't work because it was sometimes called with the window itself. Other controls didn't work because they weren't correctly parented. I fixed all that, then decided that it was dumb to keep traversing the hierarchy and changed to storing ptrs to the buttons in the struct.

That led me down a bit of a rabbit-hole and now the C API to dialog-options is gone and we have a proper class GncOptionsDialog replacing `struct GncOptionWin`.

Reply:
:+1:

Reply:
The GncOptionsDialog class is a nice touch. Much more readable and coherent than the collection of C functions.